### PR TITLE
pga: the Public Git Archive tool

### PR DESCRIPTION
**Pull request by campoy**
Simple proposal, more filters and concurrency will be added later.

This allows the user to do things like listing all of the repositories (by url) with Go.

```
$ pga list -l go
```

Or download them directly:

```
$ pga download -l go
```

This would replace `multitool`

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@campoy Please note that multitool queries and downloads GHTorrent too, so "list" and "download" options are ambiguous.

**Comment by campoy** [comment]
Why do we need to download from GHTorrent too, @vmarkovtsev?
Wouldn't it be easier to have a tool just to download Public Git Archive?

**Comment by vmarkovtsev** [comment]
Because it is used not only to fetch the dataset but also to generate it.

**Comment by vmarkovtsev** [comment]
Hence the name actually.

**Comment by vmarkovtsev** [comment]
@campoy Let's discuss how we structure the tools next week - it deserves a discussion.

**Comment by campoy** [comment]
ok @vmarkovtsev, I just wanna make sure people have it easier to download the dataset
the current tool downloads broken files and using it is quite complicated.

maybe we need to have two different tools, one for downloading the dataset and one for generating it

feel free to send me an invite for Monday morning (PST)

**Comment by rabimba** [comment]
So I wanted to try this.
Have a few queries

- How do I actually use the tool? Cloned this branch. From main directory of pga if I run go run pga.go it gives me an error with the path (I have $GOPATH set to this directory). If I change the path to cmd inside file from github link to the folder inside directory, it still gives me an error

- This will also download the siva files like multitool? Or will it download the actual repo? Will I be able to run filters on "pga" or the workflow is like previous where I worn on those siva files using engine?

**Comment by campoy** [comment]
Hey @rabimba I'll provide more documentation, and possibly binary distributions of the tool later on.

To get the binary:
- install golang: golang.org/doc/install
- clone the branch inside of $GOPATH/src/github.com/src-d/datasets
- `go install github.com/src-d/datasets/PublicGitArchive/gpa`

This will save the binary under `$GOPATH/bin`.

Then run the binary `pga list -l java` to list all the repositories containing java
Then `pga get -l java` to download the `.siva` files.

**Comment by campoy** [comment]
Adding @mcuadros just in case he has the time to review.

**Comment by campoy** [comment]
Hey @eiso I haven't seen of those comments :-/
Are you sure you actually sent them?

**Comment by eiso** [comment]
Discussed via Slack.

File: PublicGitArchive/pga/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+## Installation
+
+There are no binary distributions available yet, but we're planning on releasing them sometime soon.
+In the meanwhile you'll need to compile this tool.
+
+1. install Go (https://golang.org/doc/install).
+1. fetch the source code: `go get github.com/src-d/datasets/PublicGitArchive/pga`
+1. a new binary is now avilable under `$GOPATH/bin`: `echo "$(go env GOPATH)/bin"`
+1. add that binary to your `PATH` environment variable or move the binary to somewhere easier to find.
+1. verify the installation went well, simply run `gpa -h` and you should see some help.
```
[the remark this answers is not in the document]
Comment (campoy) [reply PublicGitArchive/pga/README.md]: done

File: PublicGitArchive/pga/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+## Utilization
+
+There are two subcommands in `gpa`: `list` and `get`.
```
[the remark this answers is not in the document]
Comment (campoy) [reply PublicGitArchive/pga/README.md]: done

File: PublicGitArchive/pga/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+### Listing repositories
+
+When you run `gpa list` two things wil happen.
```
[the remark this answers is not in the document]
Comment (campoy) [reply PublicGitArchive/pga/README.md]: done

File: PublicGitArchive/pga/README.md [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+Read below how to download repositories given the siva filenames.
+
+### Downloading siva files
+
+Simply replace `list` with `get`! You also get a couple of extra flags.
+
+- `--output path` (or `-o path`) determines under what path the siva files should be stored.
+  - if the path is a URL with schema `hdfs` HDFS will be used.
+- `--jobs n` (or `-j n`) sets the maximum number of download hapenning concurrently.
```
[the remark this answers is not in the document]
Comment (campoy) [reply PublicGitArchive/pga/README.md]: done

**Comment by vmarkovtsev** [comment]
@campoy Please never merge PRs without a formal approval of at least one of the maintainers. The process is described in https://github.com/src-d/guide/blob/master/engineering/maintainers.md The maintainers list of this repository is https://github.com/src-d/datasets/blob/master/MAINTAINERS.md

**Comment by vmarkovtsev** [comment]
Also you have violated the contribution guidelines, namely the commit message template. https://github.com/src-d/datasets/blob/master/CONTRIBUTING.md#format-of-the-commit-message

**Comment by campoy** [comment]
Hey, sorry I didn't wait for long enough before merging.

Would you want me to fix the commit messages? I'm happy to do so if you consider it necessary.

**Pull request merged**